recursion/problem-set-recursion-pt-1.md

```diff
@@ -0,0 +1,277 @@
+
+Here are the tests:
+
+```python
+def test_bunny_zero_bunnies():
+    assert bunny(0) == 0
+
+
+def test_bunny_one_bunny_has_two_ears():
+    assert bunny(1) == 2
+
+
+def test_bunny_fifty_bunnies_have_100_ears():
+    assert bunny(50) == 100
+```
+
+### !end-question
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_bunny_zero_bunnies(self):
+        self.assertEqual(bunny(0), 0)
+
+    def test_bunny_one_bunny_has_two_ears(self):
+        self.assertEqual(bunny(1), 2)
+
+    def test_bunny_fifty_bunnies_have_100_ears(self):
+        self.assertEqual(bunny(50), 100)
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def bunny(count):
+    if not count:
+        return count
+    else:
+        return 2 + bunny(count-1)
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 4 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: hhiNnl
+* title: Recursion
+* points: 3
+### !question
+
+Write a function `is_nested_parens` that accepts a string `input` of only parenthesis. It returns `True` if those parentheses are properly nested, and `False` if they are not. You may assume that no non-parenthesis characters will be passed to this function.
+
+| Example `input` | Expected Output |
+| --------------- | --------------- |
+| `"((()))"`      | `True`          |
+| `""`            | `True`          |
+| `"(())))"`      | `False`         |
+
+Here are the tests:
+
+```python
+def test_is_nested_parens():
+    input = "((()))"
+
+    assert is_nested_parens(input)
+
+
+def test_is_nested_parens_empty_str():
+    assert is_nested_parens("")
+
+
+def test_is_nested_parens_not_matching_length():
+    input = "(())))"
+
+    assert not is_nested_parens(input)
+```
+### !end-question
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_is_nested_parens(self):
+        input = "((()))"
+
+        self.assertTrue(is_nested_parens(input))
+
+    def test_is_nested_parens_empty_str(self):
+        self.assertTrue(is_nested_parens(""))
+
+    def test_is_nested_parens_not_matching_length(self):
+        input = "(())))"
+
+        self.assertFalse(is_nested_parens(input))
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def is_nested_parens(input):
+    if len(input) == 0:
+        return True
+    elif input[:1] != "(" or input[-1:] != ")":
+        return False
+    else:
+        return is_nested_parens(input[1:-1])
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### !callout-info
+
+## Challenge for `is_nested_parens`
+
+For an additional challenge, implement `is_nested_parens` without creating new strings in the process of solving the problem.
+
```

Comment:
the non-string-creating solution strikes me as needing additional params for tracking more data, which we didn't really talk about in writing recursive functions. Maybe a nudge would be helpful? I've included the type of solution I had in mind in the following comment.

```suggestion

<br />

<details>
    <summary>Hint</summary>

<br />

Sometimes we create extra helper functions that can store more information in their parameters than we might use in the main solution function.

</details>

```

Reply:
```python
def is_nested_parens(parens):
    return is_nested_helper(parens, 0, 0)
    
def is_nested_helper(parens, pos, depth):
    if pos >= len(parens):
        return True
    
    depth += 1 if parens[pos] == '(' else -1
    
    if depth < 0:
        return False
        
    return is_nested_helper(parens, pos + 1, depth)
```

Reply:
I'll add this sample solution to an instructor file, thank you!!